File: card-base/src/main/java/com/adyen/checkout/card/CardComponentProvider.java
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 23/7/2019.
 */

package com.adyen.checkout.card;

import android.app.Application;
import androidx.lifecycle.ViewModelProviders;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import android.text.TextUtils;

import com.adyen.checkout.base.ComponentAvailableCallback;
import com.adyen.checkout.base.PaymentComponentProvider;
import com.adyen.checkout.base.component.lifecycle.PaymentComponentViewModelFactory;
import com.adyen.checkout.base.model.paymentmethods.PaymentMethod;
import com.adyen.checkout.card.data.CardType;
import com.adyen.checkout.core.log.LogUtil;
import com.adyen.checkout.core.log.Logger;

import java.util.ArrayList;
import java.util.List;

public class CardComponentProvider implements PaymentComponentProvider<CardComponent, CardConfiguration> {
    private static final String TAG = LogUtil.getTag();

    @NonNull
    @Override
    public CardComponent get(@NonNull FragmentActivity activity, @NonNull PaymentMethod paymentMethod, @NonNull CardConfiguration configuration) {
        final CardConfiguration verifiedConfiguration = checkSupportedCardTypes(paymentMethod, configuration);
        final PaymentComponentViewModelFactory factory = new PaymentComponentViewModelFactory(paymentMethod, verifiedConfiguration);
        return ViewModelProviders.of(activity, factory).get(CardComponent.class);
    }

    @NonNull
    @Override
    public CardComponent get(@NonNull Fragment fragment, @NonNull PaymentMethod paymentMethod, @NonNull CardConfiguration configuration) {
        final CardConfiguration verifiedConfiguration = checkSupportedCardTypes(paymentMethod, configuration);
        final PaymentComponentViewModelFactory factory = new PaymentComponentViewModelFactory(paymentMethod, verifiedConfiguration);
        return ViewModelProviders.of(fragment, factory).get(CardComponent.class);
    }

    @Override
    public void isAvailable(
            @NonNull Application applicationContext,
            @NonNull PaymentMethod paymentMethod,
            @NonNull CardConfiguration configuration,
            @NonNull ComponentAvailableCallback<CardConfiguration> callback) {

        final boolean isPubKeyAvailable = !TextUtils.isEmpty(configuration.getPublicKey());
        callback.onAvailabilityResult(isPubKeyAvailable, paymentMethod, configuration);
    }

    /**
     * Check which set of supported cards to pass to the component.
     * Priority is: Custom -> PaymentMethod.brands -> Default
     *
     * @param paymentMethod The payment methods object that will start the component.
     * @param cardConfiguration The configuration object that will start the component.
     * @return The Configuration object with possibly adjusted values.
     */
    private CardConfiguration checkSupportedCardTypes(@NonNull PaymentMethod paymentMethod, @NonNull CardConfiguration cardConfiguration) {
        if (cardConfiguration.getSupportedCardTypes().isEmpty()) {
            final List<String> brands = paymentMethod.getBrands();

            List<CardType> supportedCardTypes = CardConfiguration.DEFAULT_SUPPORTED_CARDS_LIST;

            // Get card types from brands in PaymentMethod object
            if (brands != null && !brands.isEmpty()) {
                supportedCardTypes = new ArrayList<>();
                for (String brand : brands) {
                    final CardType brandType = CardType.getCardTypeByTxVariant(brand);
                    if (brandType != null) {
                        supportedCardTypes.add(brandType);
                    } else {
                        Logger.e(TAG, "Failed to get card type for brand: " + brand);
                    }
                }
            } else {
                Logger.d(TAG, "Falling back to DEFAULT_SUPPORTED_CARDS_LIST");
            }

            return cardConfiguration.newBuilder()
                    .setSupportedCardTypes(supportedCardTypes.toArray(new CardType[0]))
                    .build();
        }
        return cardConfiguration;
    }
}

File: card/src/main/java/com/adyen/checkout/card/CardComponentProvider.kt
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 23/7/2019.
 */
package com.adyen.checkout.card

import android.os.Bundle
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelStoreOwner
import androidx.savedstate.SavedStateRegistryOwner
import com.adyen.checkout.card.data.CardBrand
import com.adyen.checkout.card.repository.AddressRepository
import com.adyen.checkout.card.repository.BinLookupRepository
import com.adyen.checkout.components.StoredPaymentComponentProvider
import com.adyen.checkout.components.base.lifecycle.viewModelFactory
import com.adyen.checkout.components.model.paymentmethods.PaymentMethod
import com.adyen.checkout.components.model.paymentmethods.StoredPaymentMethod
import com.adyen.checkout.components.repository.PublicKeyRepository
import com.adyen.checkout.core.log.LogUtil
import com.adyen.checkout.core.log.Logger

private val TAG = LogUtil.getTag()

class CardComponentProvider : StoredPaymentComponentProvider<CardComponent, CardConfiguration> {

    override fun <T> get(
        owner: T,
        paymentMethod: PaymentMethod,
        configuration: CardConfiguration
    ): CardComponent where T : SavedStateRegistryOwner, T : ViewModelStoreOwner {
        return get(owner, owner, paymentMethod, configuration, null)
    }

    override fun get(
        savedStateRegistryOwner: SavedStateRegistryOwner,
        viewModelStoreOwner: ViewModelStoreOwner,
        paymentMethod: PaymentMethod,
        configuration: CardConfiguration,
        defaultArgs: Bundle?
    ): CardComponent {
        val verifiedConfiguration = checkSupportedCardTypes(paymentMethod, configuration)
        val binLookupRepository = BinLookupRepository()
        val publicKeyRepository = PublicKeyRepository()
        val addressDelegate = AddressDelegate(AddressRepository())
        val cardValidationMapper = CardValidationMapper()
        val factory = viewModelFactory(savedStateRegistryOwner, defaultArgs) { savedStateHandle ->
            CardComponent(
                savedStateHandle,
                NewCardDelegate(
                    paymentMethod,
                    verifiedConfiguration,
                    binLookupRepository,
                    publicKeyRepository,
                    addressDelegate,
                    cardValidationMapper
                ),
                verifiedConfiguration
            )
        }
        return ViewModelProvider(viewModelStoreOwner, factory).get(CardComponent::class.java)
    }

    override fun <T> get(
        owner: T,
        storedPaymentMethod: StoredPaymentMethod,
        configuration: CardConfiguration
    ): CardComponent where T : SavedStateRegistryOwner, T : ViewModelStoreOwner {
        return get(owner, owner, storedPaymentMethod, configuration, null)
    }

    override fun <T> get(
        owner: T,
        storedPaymentMethod: StoredPaymentMethod,
        configuration: CardConfiguration,
        key: String?
    ): CardComponent where T : SavedStateRegistryOwner, T : ViewModelStoreOwner {
        return get(owner, owner, storedPaymentMethod, configuration, null, key)
    }

    override fun get(
        savedStateRegistryOwner: SavedStateRegistryOwner,
        viewModelStoreOwner: ViewModelStoreOwner,
        storedPaymentMethod: StoredPaymentMethod,
        configuration: CardConfiguration,
        defaultArgs: Bundle?
    ): CardComponent {
        return get(savedStateRegistryOwner, viewModelStoreOwner, storedPaymentMethod, configuration, defaultArgs, null)
    }

    override fun get(
        savedStateRegistryOwner: SavedStateRegistryOwner,
        viewModelStoreOwner: ViewModelStoreOwner,
        storedPaymentMethod: StoredPaymentMethod,
        configuration: CardConfiguration,
        defaultArgs: Bundle?,
        key: String?
    ): CardComponent {
        val publicKeyRepository = PublicKeyRepository()
        val factory = viewModelFactory(savedStateRegistryOwner, defaultArgs) { savedStateHandle ->
            CardComponent(
                savedStateHandle,
                StoredCardDelegate(
                    storedPaymentMethod,
                    configuration,
                    publicKeyRepository
                ),
                configuration
            )
        }

        return if (key == null) {
            ViewModelProvider(viewModelStoreOwner, factory)[CardComponent::class.java]
        } else {
            ViewModelProvider(viewModelStoreOwner, factory)[key, CardComponent::class.java]
        }
    }

    /**
     * Check which set of supported cards to pass to the component.
     * Priority is: Custom -> PaymentMethod.brands -> Default
     *
     * @param paymentMethod The payment methods object that will start the component.
     * @param cardConfiguration The configuration object that will start the component.
     * @return The Configuration object with possibly adjusted values.
     */
    private fun checkSupportedCardTypes(paymentMethod: PaymentMethod, cardConfiguration: CardConfiguration): CardConfiguration {
        if (cardConfiguration.supportedCardBrands.isNotEmpty()) {
            return cardConfiguration
        }

        val brands = paymentMethod.brands
        var supportedCardBrands = CardConfiguration.DEFAULT_SUPPORTED_CARDS_LIST.map {
            CardBrand(it)
        }

        // Get card brands from brands in PaymentMethod object
        if (!brands.isNullOrEmpty()) {
            supportedCardBrands = arrayListOf()
            for (brand in brands) {
                val cardBrand = CardBrand(brand)
                supportedCardBrands.add(cardBrand)
            }
        } else {
            Logger.d(TAG, "Falling back to DEFAULT_SUPPORTED_CARDS_LIST")
        }
        @Suppress("SpreadOperator")
        return cardConfiguration.newBuilder()
            .setSupportedCardTypes(*supportedCardBrands.toTypedArray())
            .build()
    }
}
